### PR TITLE
Add support for Python 3.11.

### PR DESCRIPTION
**Pull request by icemac**
Including a workaround for https://github.com/python/cpython/issues/90113 which occurred just in the Python 3.11 test runs.

### CODE REVIEW

**Review by dataflake — APPROVED** [review]

File: src/zope/app/apidoc/tests.py [hunk 1]
```diff
@@ -238,14 +240,21 @@ def test_keyword_only_arguments(self):
...
+                self.assertEqual(
+                    simple_sig, "(self, mode='r', buffering=None)")
+                self.assertEqual(
+                    '(self)', getFunctionSignature(
+                        TestUtilities.test_keyword_only_arguments))
+            else:
+                self.assertEqual(simple_sig, "(mode='r', bufsize=-1)")
+                self.assertEqual(
+                    '()', getFunctionSignature(
+                        TestUtilities.test_keyword_only_arguments))
```
Comment (mgedmin) [inline src/zope/app/apidoc/tests.py:253]: AFAICT previously we had _simpleGetFunctionSignature() broken on Python 3, when called on socket.makefile(), and thus the Python 3 specific test assertions were in the `except:` branch leaving this `else:` branch aiming at Python 2.

Now you fixed `_simpleGetFunctionSignature(socket.makefile` to work on Python 3, which suggests that the `try:` block and the entire `except:` branch can be removed.  Am I right?

Comment (mgedmin) [reply src/zope/app/apidoc/tests.py:253]: Actually, things are even simpler than I thought: `_simpleGetFunctionSignature()` is only used when `inspect.signature` does not exist, i.e. on Python <= 3.2.  Why not `@unittest.skipIf(six.PY3)` the entire test?

File: src/zope/app/apidoc/utilities.py [hunk 2]
```diff
@@ -238,7 +238,11 @@ def _simpleGetFunctionSignature(func, ignore_self=False):
     _checkFunctionType(func)
 
     try:
-        args, varargs, varkw, defaults = inspect.getargspec(func)
+        if six.PY3:
+            (args, varargs, varkw, defaults, kwonlyargs, kwonlydefaults,
+             annotations) = inspect.getfullargspec(func)
```
Comment (mgedmin) [inline src/zope/app/apidoc/utilities.py:243]: If this is used only on Python 2, then we don't need to make it work on Python 3.

**Review by mgedmin — APPROVED** [review]
LGTM.

I don't fully understand the 404 page test shenanigans.

The test fixes for _simpleGetFunctionSignature() can be simplified (and probably ought to be simplified, otherwise we leave strange complications for future maintainers) since that's a fallback function used only on Python 2.7 and we don't care if it's broken on 3.11.

File: src/zope/app/apidoc/tests.py (hunk 1 above)
Comment (icemac) [reply src/zope/app/apidoc/tests.py]: Thank you for catching this issue. I adapted the code in https://github.com/zopefoundation/zope.app.apidoc/pull/38/commits/61de1109be0e8a45ea6dbc2a64f1d4bf87e47940 but Python 2 is no longer supported by tox 4, so I'll drop Python 2 support.

File: src/zope/app/apidoc/utilities.py (hunk 2 above)
Comment (icemac) [reply src/zope/app/apidoc/utilities.py]: See https://github.com/zopefoundation/zope.app.apidoc/pull/38/commits/61de1109be0e8a45ea6dbc2a64f1d4bf87e47940

**Comment by icemac** [comment]
Thank you for reviewing this PR. 😃

**Pull request merged**